### ISSUE TITLE
How should devDependencies work?

### ISSUE THREAD

**Issue by Profpatsch** [Profpatsch/yarn2nix#5]
There’s a [TODO in the code](https://github.com/Profpatsch/yarn2nix/blob/9ecaf3e7a89328ed206eb4ed7143bb8c990543ad/src/Distribution/Nixpkgs/Nodejs/FromPackage.hs#L30) and apparently in [some cases `devDependencies` is actually needed](https://github.com/mayflower/yarn2nix/commit/10eb1e2fbb128a693dfc492852fd00d2487ce914)?

What’s people’s thoughts about use-cases and how they should be integrated?

**Comment by calbrecht** [comment]
devDependencies are used while developing the packaged application/library. Also on build time e. g. for packing bundles with webpack or transpiling different js dialects into another js dialect.

According to https://yarnpkg.com/en/docs/dependency-types

These are your development dependencies. Dependencies that you need at some point in the development workflow but not while running your code (e.g. Babel or Flow).

**Comment by Pauan** [comment]
This is my understanding of `devDependencies`:

Before publishing code to npm, there is sometimes a need to run an additional pre-processing step. Some examples are minifying the code, transpiling from ES6 to ES5, bundling with Webpack, running unit tests, building documentation, etc.

So the package will put `uglify-js` or `babel-core` or whatever into their `devDependencies`, and then they add a `prepare` script to their `package.json`:

```json
{
  "devDependencies": {
    "uglify-js": "^3.3.11"
  },
  "scripts": {
    "prepare": "uglifyjs foo.js --compress --mangle --output foo.min.js"
  }
}
```

Now, while developing the package, the package author can run `npm install`, which will download the `devDependencies`, and then run the `prepare` script.

When the author is done developing the package, they can run `npm publish` to publish the package to npm. The end result is that the npm package contains the *final pre-processed code*.

When using that npm package as a dependency in another npm package, it will ***not*** download `devDependencies`, because there is no need to do so: the npm package already contains the pre-processed files.

In essence, `devDependencies` serves the same purpose as `buildInputs` in Nix, and `dependencies` serves the same purpose as `propagatedBuildInputs`.

Some caveats: as explained earlier, when using a normal npm dependency it will ***not*** download `devDependencies`, and it will ***not*** run `prepare` for that package.

In addition, when using a *local* dependency, such as `"foo": "file:./path/to/package"` it will ***not*** download `devDependencies` or run the `prepare` script for `foo`.

***However*** when using a *git repository* as a dependency, it ***will*** download `devDependencies` and run the `prepare` script.

It also downloads `devDependencies` and runs the `prepare` script when using `npm link`.

Here is the official npm documentation for this:

https://docs.npmjs.com/files/package.json#devdependencies
https://docs.npmjs.com/misc/scripts#use-cases
https://github.com/npm/npm/releases/tag/v5.0.0

It seems the current behavior is a bit buggy, so it might change in the future:

https://github.com/npm/npm/issues/19684
https://github.com/npm/npm/issues/18132

----

As far as `yarn2nix` is concerned, that implies it will need to download the `devDependencies` (and run the `prepare` script) when using git dependencies, but it doesn't need to download `devDependencies` for regular npm dependencies or local dependencies.

**Comment by calbrecht** [comment]
https://github.com/webpack/webpack/issues/520

**Comment by Profpatsch** [comment]
Maybe it’s sensible to state my assumption: I did not assume `yarn2nix` would be used to develop packages or download dependencies for development, in order to reduce complexity. `yarn` does a great job at this.

So the only question we need to ask ourselves is whether `devDependencies` are needed when packaging certain applications.


> As far as yarn2nix is concerned, that implies it will need to download the devDependencies (and run the prepare script) when using git dependencies, but it doesn't need to download devDependencies for regular npm dependencies or local dependencies.

That’s a great first heuristic, so one idea could be “enable `devDependencies` for git sources by default”.

What we definitely need is an easy way to override whether to enable `devDependencies` for specific packages, maybe even keep a global database to make it “just work” by default for often-used dependencies.

But first we should get a good understanding of the problem space, so *please* open issues for each failing package you encounter, that we can solve these failures one by one.

Edit: In certain cases, patching the `package.json` might be the way to go, because [lots of npm authors seem to be confused about the distinction](https://github.com/webpack/webpack/issues/520#issuecomment-174011824).

**Comment by Pauan** [comment]
> Maybe it’s sensible to state my assumption: I did not assume yarn2nix would be used to develop packages or download dependencies for development, in order to reduce complexity. yarn does a great job at this.

Yup, I figured that. But I wanted to thoroughly explain how npm / yarn handle `devDependencies`, so we can ensure that `yarn2nix` has the same behavior.

It's hard to find documentation about this kind of stuff, and npm has a lot of corner cases, exceptions, and weird behavior, so it's nice to know what the full expected behavior is.

> Edit: In certain cases, patching the package.json might be the way to go, because lots of npm authors seem to be confused about the distinction.

That definitely seems like a good idea in general (not just for `devDependencies`)! It might also help solve issues like https://github.com/Profpatsch/yarn2nix/issues/6 (assuming we can patch the `package.json` to move the bad packages into `devDependencies`)

**Comment by Profpatsch** [comment]
> Yup, I figured that. But I wanted to thoroughly explain how npm / yarn handle devDependencies, so we can ensure that yarn2nix has the same behavior.
> 
> It's hard to find documentation about this kind of stuff, and npm has a lot of corner cases, exceptions, and weird behavior, so it's nice to know what the full expected behavior is.

Thanks for the link collection, having references to all this really helps (I know from experience how scattered the npm documentation is, having only [found out about scopes](https://github.com/Profpatsch/yarn2nix/issues/4) after hitting an extremely strange bug.

**Comment by Profpatsch** [comment]
Okay, since [last commit](https://github.com/Profpatsch/yarn2nix/commit/ecb4b21988711b51e6be11e70a2ba5f322e6a83c) package templates (generated with `--template`) always include the `devDependencies` of that outermost package, since most projects seem to rely on them for build (but they are normally not needed if a package comes from an npm registry).

I have already found the next problem regarding dependencies, though: The `browserify-incremental` packages [has `browserify` only in its `peerDependencies` and `devDependencies`](https://github.com/jsdf/browserify-incremental/blob/master/package.json#L22), even though it clearly always needs `browserify` at runtime. I suppose the developer expects the package to be only used at development time (and that every devDependency of every package exists in that case).

**Comment by utdemir** [comment]
Same issue happens on https://github.com/jantimon/html-webpack-plugin, where it only requires `webpack` in `devDependnecies`, however it clearly needs it at "runtime" (it's only meant to be used when building an application via webpack). However, currently using `yarn2nix` to build an application using `html-webpack-plugin` fails with something like:

```
Error: Cannot find module 'webpack/lib/node/NodeTemplatePlugin'
Require stack:
- /nix/store/k1nl5k17z2dfwm0y1nj95fv7ggi8as1d-html-webpack-plugin-4.5.0/lib/child-compiler.js
- /nix/store/k1nl5k17z2dfwm0y1nj95fv7ggi8as1d-html-webpack-plugin-4.5.0/lib/cached-child-compiler.js
- /nix/store/k1nl5k17z2dfwm0y1nj95fv7ggi8as1d-html-webpack-plugin-4.5.0/index.js
- /nix/store/yn27bl44d72mjl1vkhvw86725zsgjd9b-react-scripts-4.0.1/config/webpack.config.js
- /nix/store/yn27bl44d72mjl1vkhvw86725zsgjd9b-react-scripts-4.0.1/scripts/build.js
```

Which is a bit unfortunate, since as far as I can see `create-react-app` is widely used, which has `html-webpack-plugin` as a transitive depedendency.
